src/core/MIGRATION.md

```diff
@@ -653,20 +671,66 @@ At the very least, any plugin exposing an extension point should do so with firs
+- A function that queries Elasticsearch, but does not take an ES Client as a parameter.
+- A React component that reads `uiSettings` from a variable defined in the module.
+- A function that allows other plugins to register values that get pushed into an array defined internally to the module.
+
+#### So what happens if I don't listen to you?
+
+One of the outcomes of producing a stable core API in Kibana is that a single plugin version should work across all installations of the same major version of Kibana. To enable this, we are planning to use "immutable bundles" that are built once, and compatible with every minor version of that Kibana major version.
+
+- Any code you export to other plugins will get copied into their bundles, so if a plugin is built for 8.1 and is running on Kibana 8.2, any modules it imported that changed will not be updated in that plugin.
+- When a plugin is disabled, its static exports will still be importable by any other plugin. This can result in undesirable effects where it _appears_ like a plugin is enabled when it is not. In the worst case, it can result in an unexpected user experience where features that should have been disabled are not.
+- Exported modules will be copied into plugins that import them. Any state your plugin populates to a exported module will not be the same instance of the module that dependent plugins are using. Effectively, impure exported modules simply won't work and your plugin will not have access to the copy of the module in other plugins.
+
+#### Concrete Example
+
+Ok, you've decided you want to export static code from your plugin, how do you do it? In principle, the New Platform only considers values exported from `my_plugin/public` and `my_plugin/server` to be stable. Our tooling will enforce that these APIs do not break between major versions (in the future). All code shared among plugins should be exported in these modules like so:
+
+```ts
+// my_plugin/public/index.ts
+export { MyPureComponent } from './components';
+
+// regular plugin export used by core to initialize your plugin
+export const plugin = ...;
+```
+
+These can then be imported using relative paths from other plugins:
+
+```ts
+// my_other_plugin/public/components/my_app.ts
+import { MyPureComponent } from '../my_plugin/public';
+```
 
-Any code not exported via the index of either the `server` or `public` directories should never be imported outside that plugin as it should be considered unstable and subject to change at any time.
+If you have code that should be available to other plugins on both the client and server, you can export the same value in both `my_plugin/public` and `my_plugin/server`. Plugins should not ever import code from deeply inside another plugin (eg. `my_plugin/public/components`) or from other top-level directories (eg. `my_plugin/common/constants`) as these are not checked for breaking changes and are considered unstable and subject to change at any time.
```

Comment:
probably worth adding:
you **can** use `my_plugin/common` folder inside of your plugin, just don't export outside of the your plugin boundaries.